File: data-structure-and-algorithm/jian-zhi-offer/0012.Route_In_Matrix/RouteInMatrix.java
public class RouteInMatrix {

	public boolean hasPath(char[] matrix, int rows, int cols, char[] str) {
		boolean[] visited = new boolean[rows * cols];
		for (int i = 0; i < rows; i++) {
			for (int j = 0; j < cols; j++) {
				if (dfs(matrix, rows, cols, i, j, visited, 0, str)) {
					return true;
				}
			}
		}
		return false;
	}

	public boolean dfs(char[] matrix, int rows, int cols, int row, int col, boolean[] visited,
					   int k, char[] str) {
		boolean result = false;
		if (row >= 0 && row < rows && col >= 0 && col < cols && !visited[row * cols + col] && matrix[row * cols + col] == str[k]) {
			visited[row * cols + col] = true;
			k++;
			if (k == str.length) {
				return true;
			}
			if(k==10){
				System.out.println("hello");
			}
			result = dfs(matrix, rows, cols, row + 1, col, visited, k, str) ||
				dfs(matrix, rows, cols, row - 1, col, visited, k, str) ||
				dfs(matrix, rows, cols, row, col + 1, visited, k, str) ||
				dfs(matrix, rows, cols, row, col - 1, visited, k, str);
			if(!result){
				k--;
				visited[row * cols + col] = false;
			}
		}
		return result;
	}

	public static void main(String[] args) {
		RouteInMatrix test = new RouteInMatrix();
//		System.out.println(test.hasPath("ABCESFCSADEE".toCharArray(), 3, 4, "ABCCED".toCharArray()));
		System.out.println(test.hasPath("ABCEHJIGSFCSLOPQADEEMNOEADIDEJFMVCEIFGGS".toCharArray(),
			5, 8, "SGGFIECVAASABCEHJIGQEM".toCharArray()));
	}
}

File: data-structure-and-algorithm/jian-zhi-offer/0012.Route-In-Matrix/RouteInMatrix.py
from typing import List


class RouteInMatrix:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(i: int, j: int, k: int):
            if not 0 <= i < len(board) or not 0 <= j < len(board[0]) or not board[i][j] == word[k] :
                return False
            if k == len(word) - 1:
                return True
            tmp, board[i][j] = board[i][j], '*'
            result = dfs(i + 1, j, k + 1) or dfs(i, j + 1, k + 1) or dfs(i - 1, j, k + 1) or dfs(i, j - 1, k + 1)
            board[i][j] = tmp
            return result
        for i in range(len(board)):
            for j in range(len(board[0])):
                if(dfs(i, j, 0)):
                    return True
        return False


if __name__ == "__main__":
    test = RouteInMatrix()
    board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]]
    word = "ABCCED"
    print(test.exist(board, word))